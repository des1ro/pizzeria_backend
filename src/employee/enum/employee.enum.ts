export enum EmployeeRole {
  Cheff = "Cheff",
  Waiter = "Waiter",
}
