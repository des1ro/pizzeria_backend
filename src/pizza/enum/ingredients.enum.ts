export enum Ingredients {
  tomato = "Tomato",
  peperoni = "Peperoni",
  Cheese = "Cheese",
  Eggs = "Eggs",
  Salad = "Salad",
  Pineapple = "Pineapple",
}
