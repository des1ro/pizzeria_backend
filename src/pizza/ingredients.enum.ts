export enum Ingredients {
  tomato,
  peperoni,
  cheese,
  eggs,
  salad,
  pineapple,
}
