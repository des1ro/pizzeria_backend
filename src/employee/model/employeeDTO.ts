import { EmployeeRole } from "../enum/employee.enum";

export class EmployeeDTO {
  constructor(readonly name: string, readonly role: EmployeeRole) {}
}
