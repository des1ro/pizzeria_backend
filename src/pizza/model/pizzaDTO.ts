import { Ingredients } from "../enum/ingredients.enum";

export class PizzaDTO {
  constructor(
    readonly name: string,
    readonly price: number,
    readonly ingredients: Ingredients[]
  ) {}
}
