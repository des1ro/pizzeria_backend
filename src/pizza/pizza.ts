import { Ingredients } from "./ingredients.enum";

export class Pizza {
  constructor(
    readonly name: string,
    readonly value: number,
    readonly ingredients: Ingredients[]
  ) {}
}
