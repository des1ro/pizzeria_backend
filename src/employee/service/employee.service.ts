import { EmployeeRole } from "../enum/employee.enum";
import { EmployeeDTO } from "../model/employeeDTO";
import { EmployeeServiceError } from "../error/employee.exception";

export class EmployeeService {
  private readonly takenCheffs = new Set<EmployeeDTO>();
  constructor(
    private readonly availableCheffs = new Set<EmployeeDTO>(),
    readonly availableWaiters = new Set<EmployeeDTO>()
  ) {}
  addCheffToCrew(name: string): void {
    const cheff = new EmployeeDTO(name, EmployeeRole.Cheff);
    this.availableCheffs.add(cheff);
  }
  getCheffToOrder(): EmployeeDTO {
    const cheff = Array.from(this.availableCheffs).reverse().pop();
    if (cheff) {
      this.takenCheffs.add(cheff);
      this.availableCheffs.delete(cheff);
      return cheff;
    }
    throw new EmployeeServiceError({
      name: "EMPLOYEE_ERROR",
      message: "There isn't free cheffs",
    });
  }
  isCheffAvailable(): boolean {
    return this.availableCheffs.size > 0;
  }
  relievedCheff(cheff: EmployeeDTO) {
    if (this.takenCheffs.delete(cheff)) {
      this.availableCheffs.add(cheff);
      return;
    }
    throw new EmployeeServiceError({
      name: "EMPLOYEE_ERROR",
      message: "This cheff wasn't taken",
    });
  }
}
