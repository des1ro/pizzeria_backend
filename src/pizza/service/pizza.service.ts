import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { Ingredients } from "../enum/ingredients.enum";
import { PizzaDTO } from "../model/pizzaDTO";

export class PizzaService {
  constructor(
    private readonly ingredients = new Map<Ingredients, number>(),
    private readonly pizzaList = new Set<PizzaDTO>()
  ) {}
  addIngredient(ingredient: Ingredients, amount = 1): void {
    if (this.ingredients.has(ingredient)) {
      const oldAmount = this.ingredients.get(ingredient) || 0;
      this.ingredients.set(ingredient, oldAmount + amount);
      return;
    }
    this.ingredients.set(ingredient, amount);
  }
  addPizza(pizza: PizzaDTO): void {
    this.pizzaList.add(pizza);
  }
  delatePizza(pizzaName: string): void {
    const pizza = this.findPizzaByName(pizzaName);
    this.pizzaList.delete(pizza);
  }
  getIngredientsMapOfOrder(pizzaNames: string[]) {
    const map = new Map(this.ingredients);
    for (const pizzaName of pizzaNames) {
      const pizza = this.findPizzaByName(pizzaName);
      if (pizza) {
        for (const ingredient of pizza.ingredients) {
          const value = map.get(ingredient);
          if (value === undefined || value <= 0) {
            throw new PizzeriaError({
              name: "PIZZA_SERVICE_ERROR",
              message: "Insufficient amount of ingredients",
            });
          }
          map.set(ingredient, value - 1);
        }
      }
    }
    return map;
  }
  updateIngredientsAfterOrder(map: Map<Ingredients, number>): void {
    this.ingredients.clear();
    map.forEach((value, key) => {
      this.ingredients.set(key, value);
    });
  }
  getPizzaFromNames(pizzaNames: string[]): PizzaDTO[] {
    const pizzas: PizzaDTO[] = [];
    pizzaNames.forEach((pizzaName) => {
      const pizza = this.findPizzaByName(pizzaName);
      pizzas.push(pizza);
    });
    return pizzas;
  }
  private findPizzaByName(pizzaName: string): PizzaDTO {
    const pizza = Array.from(this.pizzaList).find(
      (pizza) => pizza.name === pizzaName
    );
    if (pizza) {
      return pizza;
    }
    throw new PizzeriaError({
      name: "PIZZA_SERVICE_ERROR",
      message: "Pizza isn't on the list",
    });
  }
}
