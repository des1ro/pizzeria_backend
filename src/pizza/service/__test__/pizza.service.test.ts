import { PizzeriaError } from "../../../pizzeria/error/pizzeria.exceptions";
import { Ingredients } from "../../enum/ingredients.enum";
import { PizzaServiceError } from "../../error/pizza.exceptions";
import { PizzaDTO } from "../../model/pizzaDTO";
import { PizzaService } from "../pizza.service";

describe("PizzaService test suite", () => {
  let objectUnderTest: PizzaService;
  let mockedPizzaOne: PizzaDTO;
  let mockedPizzaNameOne: string;
  let mockedPizzaNameTwo: string;
  let ingredientsMap: Map<Ingredients, number>;
  let ingredientsOne: Ingredients[];
  let ingredientsTwo: Ingredients[];
  const addIngredents = () => {
    objectUnderTest.addIngredient(Ingredients.Cheese, 0);
    objectUnderTest.addIngredient(Ingredients.peperoni, 2);
    objectUnderTest.addIngredient(Ingredients.Pineapple, 2);
    objectUnderTest.addIngredient(Ingredients.tomato, 2);
  };
  beforeEach(() => {
    objectUnderTest = new PizzaService();
  });
  beforeAll(() => {
    mockedPizzaNameOne = "test pizza one";
    mockedPizzaNameTwo = "test pizza two";
    ingredientsMap = new Map<Ingredients, number>();
    ingredientsOne = [Ingredients.Cheese, Ingredients.Pineapple];
    ingredientsTwo = [Ingredients.peperoni, Ingredients.tomato];
  });
  it("Should add new ingredient to map", () => {
    //Given
    const mockedIngredient = Ingredients.Cheese;
    //When
    objectUnderTest.addIngredient(mockedIngredient);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should update amount of ingredient if its already in map", () => {
    //Given
    const mockedIngredient = Ingredients.Eggs;
    //When
    objectUnderTest.addIngredient(mockedIngredient);
    objectUnderTest.addIngredient(mockedIngredient, 3);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should add pizza to the set", () => {
    //Given
    mockedPizzaOne = new PizzaDTO("test pizza", 30, []);
    //When
    objectUnderTest.addPizza(mockedPizzaOne);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should throw error if there isn't enough ingredients in stock", () => {
    //Given
    addIngredents();
    mockedPizzaOne = new PizzaDTO(mockedPizzaNameOne, 30, ingredientsOne);
    const mockedPizzaTwo = new PizzaDTO(mockedPizzaNameTwo, 30, ingredientsTwo);
    //When
    objectUnderTest.addPizza(mockedPizzaOne);
    objectUnderTest.addPizza(mockedPizzaTwo);
    //Then
    expect(() =>
      objectUnderTest.updateIngredientsAfterOrder([
        mockedPizzaOne,
        mockedPizzaTwo,
      ])
    ).toThrow(PizzaServiceError);
  });
  it("Should restore ingredients in service on error", () => {
    //Given
    addIngredents();
    mockedPizzaOne = new PizzaDTO(mockedPizzaNameOne, 30, ingredientsOne);
    const mockedPizzaTwo = new PizzaDTO(mockedPizzaNameTwo, 30, ingredientsTwo);
    //When
    objectUnderTest.addPizza(mockedPizzaOne);
    objectUnderTest.addPizza(mockedPizzaTwo);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
    expect(() =>
      objectUnderTest.updateIngredientsAfterOrder([
        mockedPizzaOne,
        mockedPizzaTwo,
      ])
    ).toThrow(PizzaServiceError);
  });
});
