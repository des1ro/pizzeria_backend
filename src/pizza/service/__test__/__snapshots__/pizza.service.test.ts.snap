// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PizzaService test suite Should add new ingredient to map 1`] = `
PizzaService {
  "ingredients": Map {
    "Cheese" => 1,
  },
  "pizzaList": Set {},
}
`;

exports[`PizzaService test suite Should add pizza to the set 1`] = `
PizzaService {
  "ingredients": Map {},
  "pizzaList": Set {
    PizzaDTO {
      "ingredients": [],
      "name": "test pizza",
      "price": 30,
    },
  },
}
`;

exports[`PizzaService test suite Should restore ingredients in service on error 1`] = `
PizzaService {
  "ingredients": Map {
    "Cheese" => 0,
    "Peperoni" => 2,
    "Pineapple" => 2,
    "Tomato" => 2,
  },
  "pizzaList": Set {
    PizzaDTO {
      "ingredients": [
        "Cheese",
        "Pineapple",
      ],
      "name": "test pizza one",
      "price": 30,
    },
    PizzaDTO {
      "ingredients": [
        "Peperoni",
        "Tomato",
      ],
      "name": "test pizza two",
      "price": 30,
    },
  },
}
`;

exports[`PizzaService test suite Should update amount of ingredient if its already in map 1`] = `
PizzaService {
  "ingredients": Map {
    "Eggs" => 4,
  },
  "pizzaList": Set {},
}
`;
